activerecord/lib/active_record/associations/join_dependency/join_association.rb

```diff
@@ -63,11 +63,17 @@ def join_constraints(foreign_table, foreign_klass, join_type, alias_tracker)
                 end
               end
 
-              joins << join_type.new(table, Arel::Nodes::On.new(nodes))
-
-              if others && !others.empty?
-                joins.concat arel.join_sources
-                append_constraints(connection, joins.last, others)
+              if join_using_subquery?(scope)
+                arel.constraints.delete_at(0)
+                arel.constraints.unshift(*others)
```

Comment:
`#join_scope` returns an `ActiveRecord::Relation`. It would greatly simplify things if `#join_scope` returned `[ActiveRecord::Relation, Arel::Nodes::On]`.